Add target size enhanced rule

WilcoFiers opened the pull request:
Need for Call for Review: 2 weeks

---

## How to Review And Approve

- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines.
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.
- Make sure to also review the proposed Call for Review period. In case of disagreement, the longer period wins.

WilcoFiers committed:
Add target size enhanced rule

Jym77 committed:
Rework a bit normative parts

Jym77 committed:
Merge branch 'develop' into target-size-enhanced

Jym77 committed:
Merge branch 'develop' into target-size-enhanced

Jym77 committed:
Refine rule a bit

Jym77 committed:
Polish draft a bit

Jym77 committed:
more work

Jym77 committed:
Merge branch 'develop' into target-size-enhanced

Jym77 committed:
Attempt to write barebone defs

dd8 commented:
The lack of a CSS hit testing spec is listed as an open issue:
https://github.com/w3c/csswg-drafts/issues/2325

Hit testing doesn't seem fully interoperable. - though things may have improved since this was written:
https://lists.w3.org/Archives/Public/www-style/2010Aug/0407.html

There's a W3 Wiki article here detailing the hit testing in IE:
https://www.w3.org/wiki/Hit_Testing

Here's the code for handling touch events in Chromium:
https://github.com/chromium/chromium/blob/be0d3937d6ccc1a4e0d27315d2e827151850d3fa/android_webview/renderer/aw_render_frame_ext.cc#L233
(so Chrome actually does hit testing on rects, rather than the circle used in WCAG 2.2 2.5.8)

Jym77 committed:
Text spacing rewrite (#1923)

* Add new letter-spacing rule and deprecate old one

* Add new word-spacing rule and deprecate old one

* Clean up assumptions

* Clean up

* Clean up

* Add new line-height rule and deprecate old one

* Replace old letter spacing version rather than deprecating it

* Replace old line height version rather than deprecating it

* Replace old word spacing version rather than deprecating it

* Target text nodes

* Improve background note

* Apply suggestion from review

* Clean up

* Target text nodes rather than their parents

* Target text nodes rather than their parents

* Add missing reference

* Update example

* Apply to parent of text nodes, not text nodes

* Apply suggestions from code review

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* Typos

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* Typos

---------

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

Jym77 committed:
"Meta viewport allows for zoom" (b4f0c3): Explicit meaning of 'has' (#1994)

* Explicit meaning of 'has'

* Improve expectation and examples

* Typo

* Improve algorithm description

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* Rephrase expectations

* Streamline Applicability

* Typo

* Simplify expectations

---------

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

WilcoFiers committed:
Map Empty-heading rule to ARIA instead of WCAG (#2120)

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

WilcoFiers committed:
Deprecate 4.1.1 rules (#2117)

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

Jym77 committed:
Some more notes

WilcoFiers committed:
Fix markup issue in Iframe has name rule (#2116)

* Fix markup issue in Iframe has name rule

* Update _rules/iframe-non-empty-accessible-name-cae760.md

WilcoFiers committed:
Stop linking to proposed rules (#2108)

* Stop linking to proposed rules

* Apply suggestions from code review

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

---------

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

WilcoFiers committed:
Fix image-button accessibility support editorial oversight (#2106)

Jym77 committed:
"Iframe not with interactive element" [akn7bn]: handle inert iframe (#2125)

* Update programmatically-determined-link-context.md

* Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

* Update programmatically-determined-link-context.md

* Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

* Create modal-dialog

* Rename modal-dialog to modal-dialog.md

* Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

* Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

* Apply suggestions from code review

Co-authored-by: Dan Tripp <113939352+dan-tripp-siteimprove@users.noreply.github.com>

* Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

* Update and rename modal-dialog.md to inert.md

* Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

* Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

* Update spelling-ignore.yml

* Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

* Update inert.md

* Apply suggestions from code review

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>
Co-authored-by: Dan Tripp <113939352+dan-tripp-siteimprove@users.noreply.github.com>

* Update iframe-not-focusable-has-no-interactive-content-akn7bn.md

* Update inert.md

* Apply suggestions from code review

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

* AGWG Updates (#2067)

* Add (alt="") for clarity on empty alt

* Resolve focus visible feedback

* Tweak contrast rules

* Tweak page title descriptive

* Fix tests

* Apply suggestions from code review

Co-authored-by: Trevor R. Bostic <32486143+tbostic32@users.noreply.github.com>

---------

Co-authored-by: Trevor R. Bostic <32486143+tbostic32@users.noreply.github.com>

* [cae760] Frame has non-empty accessible name (#2034)

* First pass in response to Feb 16 TF meeting

* typo

* Update _rules/iframe-non-empty-accessible-name-cae760.md

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

* Update _rules/iframe-non-empty-accessible-name-cae760.md

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

* Update _rules/iframe-non-empty-accessible-name-cae760.md

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

* Update _rules/iframe-non-empty-accessible-name-cae760.md

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

* Update _rules/iframe-non-empty-accessible-name-cae760.md

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

* Move note about frame to background

* Set height for frame

* Test wants alphabetical contributors

* Update _rules/iframe-non-empty-accessible-name-cae760.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* Update _rules/iframe-non-empty-accessible-name-cae760.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* Update _rules/iframe-non-empty-accessible-name-cae760.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* Move note to background

---------

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>
Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* Updating glossary definition. (#2069)

* Bump yaml and zx (#2056)

* Bump yaml and zx

Bumps [yaml](https://github.com/eemeli/yaml) to 2.2.2 and updates ancestor dependency [zx](https://github.com/google/zx). These dependencies need to be updated together.


Updates `yaml` from 1.10.2 to 2.2.2
- [Release notes](https://github.com/eemeli/yaml/releases)
- [Commits](https://github.com/eemeli/yaml/compare/v1.10.2...v2.2.2)

Updates `zx` from 5.3.0 to 7.2.1
- [Release notes](https://github.com/google/zx/releases)
- [Commits](https://github.com/google/zx/compare/5.3.0...7.2.1)

---
updated-dependencies:
- dependency-name: yaml
  dependency-type: indirect
- dependency-name: zx
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>

* Trigger CLA?

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* new rule: ARIA required ID references exist (#2041)

* new rule: ARIA required ID references exist

* Address review comments

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* Apply suggestions from code review

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* Apply suggestions from code review

Co-authored-by: Tom Brunet <thbrunet@us.ibm.com>
Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* Update _rules/aria-required-id-references-in6db8.md

* Apply suggestions from code review

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

---------

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>
Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>
Co-authored-by: Tom Brunet <thbrunet@us.ibm.com>

* scrollable element: clarify the title (#2083)

* UpdateTableHeaderRule (#2074)

* UpdateTableHeaderRule

* Update table-header-cell-has-assigned-cells-d0f69e.md

* trigger test

* Update _rules/table-header-cell-has-assigned-cells-d0f69e.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

---------

Co-authored-by: Wilco Fiers <wilco.fiers@deque.com>
Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* Adds apostrophe to mark the possessive form (#2080)

Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>

* Focus visible rule: Fix typo (#2082)

* Contrast rules: Tweak background text (#2090)

* Tweak name / description of Scrollable element keyboard (#2092)

* Deprecate HTML page lang and xml:lang attributes have matching values (#2086)

* Deprecate HTML page lang and xml:lang attributes have matching values

* Apply suggestions from code review

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

---------

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* Rephrase Applicability (#2079)

* Rename file (#2078)

* Move secondary requirement texts out of the background (#2060)

* Move secondary requirement texts out of the background

* Apply suggestions from code review

* fix test

* Fix failing test

* Secondary reqs on ARIA rules

* Update all secondary requirements

* Typos

* Fix failing tests

* Update _rules/link-non-empty-accessible-name-c487ae.md

Co-authored-by: Trevor R. Bostic <32486143+tbostic32@users.noreply.github.com>

* Tweaked the language some more

* Update rule design info for secondary requirements

* Fix tests

* Apply suggestions from code review

Co-authored-by: Kathy Eng <kengdoj@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

---------

Co-authored-by: Trevor R. Bostic <32486143+tbostic32@users.noreply.github.com>
Co-authored-by: Kathy Eng <kengdoj@users.noreply.github.com>
Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* fix test on secondary requirements (#2102)

* fix test on secondary requirements

* More assertions

* Update _rules/aria-required-id-references-in6db8.md

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

---------

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* Update dependencies (including act-tools) (#2103)

* fix test-assets not getting built right (#2104)

* Update element-lang-valid-de46e4.md (#2100)

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* fix the approve-rule action (#2105)

* Remove outdated accsupport note (#2111)

* "Element with lang attribute has valid language tag" [de46e4]: Updated Failed Examples 4 and 5 to reflect Applicability (#2094)

* Update element-lang-valid-de46e4.md

Updated Failed examples 4 and 5 to reflect applicability

* Apply suggestions from code review

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* Update _rules/element-lang-valid-de46e4.md

Co-authored-by: Dan Tripp <113939352+dan-tripp-siteimprove@users.noreply.github.com>

---------

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>
Co-authored-by: Dan Tripp <113939352+dan-tripp-siteimprove@users.noreply.github.com>
Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* Text spacing rewrite (#1923)

* Add new letter-spacing rule and deprecate old one

* Add new word-spacing rule and deprecate old one

* Clean up assumptions

* Clean up

* Clean up

* Add new line-height rule and deprecate old one

* Replace old letter spacing version rather than deprecating it

* Replace old line height version rather than deprecating it

* Replace old word spacing version rather than deprecating it

* Target text nodes

* Improve background note

* Apply suggestion from review

* Clean up

* Target text nodes rather than their parents

* Target text nodes rather than their parents

* Add missing reference

* Update example

* Apply to parent of text nodes, not text nodes

* Apply suggestions from code review

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* Typos

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* Typos

---------

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* "Meta viewport allows for zoom" (b4f0c3): Explicit meaning of 'has' (#1994)

* Explicit meaning of 'has'

* Improve expectation and examples

* Typo

* Improve algorithm description

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* Rephrase expectations

* Streamline Applicability

* Typo

* Simplify expectations

---------

Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>

* Map Empty-heading rule to ARIA instead of WCAG (#2120)

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

* Deprecate 4.1.1 rules (#2117)

Co-authored-by: Jean-Yves Moyen <jym@siteimprove.com>

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: giacomo-petri <106394951+giacomo-petri@users.noreply.github.com>
Co-authored-by: Carlos Duarte <caduarte@campus.ul.pt>
Co-authored-by: Dan Tripp <113939352+dan-tripp-siteimprove@users.noreply.github.com>
Co-authored-by: Wilco Fiers <WilcoFiers@users.noreply.github.com>
Co-authored-by: Trevor R. Bostic <32486143+tbostic32@users.noreply.github.com>
Co-authored-by: Tom Brunet <thbrunet@us.ibm.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: HelenBurge <41951878+HelenBurge@users.noreply.github.com>
Co-authored-by: Wilco Fiers <wilco.fiers@deque.com>
Co-authored-by: Daniel Montalvo <49305434+daniel-montalvo@users.noreply.github.com>
Co-authored-by: Kathy Eng <kengdoj@users.noreply.github.com>

Jym77 committed:
Add much more examples

Jym77 committed:
Add much more examples

Jym77 committed:
Merge branch 'develop' into target-size-enhanced

Jym77 committed:
Add more examples

Jym77 committed:
Merge branch 'develop' into target-size-enhanced

Jym77 committed:
Revert definition of clickable area

Jym77 committed:
Clean up

Jym77 commented:
@dd8 Been trying another direction with the rule. See also https://w3ccommunity.slack.com/archives/C01167KLA73/p1700745811381929 for details.
I'd appreciate if you could get a look at the current state and test cases. You have both the technical expertise and tool insight to figure out whether this is going anywhere.

dd8 commented:
My email doesn't seem to have access to the Slack, but I've had a quick look at the changes and have a few comments. If somebody can give me access to the Slack channel I'll take a look at that as well.

I think excluding `area` elements is a good idea since they're often essential, because they're tied to the image associated with the image map (e.g.bio links on a group photo, places on a satellite photo or map). 

Given the lack of spec for clickable areas I would probably restrict the applicability to the simplest cases initially (then introduce more complex clickable areas in future versions of the rule). By simple I mean:

- HTML elements using the CSS box model (no SVG, canvas etc)
- exclude elements that have been transformed (there have been browser interop issues with CSS transforms and clickable areas for shapes like this https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/matrix)

The current applicability of 'widget' role *and* focusable looks wrong to me - it's a common a11y problem to have clickable widgets you can't tab to because they're not focusable.

The trickiest problem is knowing what's clickable. It's easy for built in widgets but not for DIVs with a 'click' event. Even if you can find which elements have a 'click' event registered there are some frameworks which put a generic 'click' event on *every element* which dispatches to handlers the author registers with the framework (e.g. the framework gives every `div` a `click` handler but the author only registers framework handlers for `div.button`, so only those trigger actions)

You may also got something like this where the clickable area is smaller than the widget role area:

```
<div role="button' style='padding:4em;'><div onclick='fireButton()'>Click me</div></div>
```

Jym77 commented:
> My email doesn't seem to have access to the Slack, but I've had a quick look at the changes and have a few comments. If somebody can give me access to the Slack channel I'll take a look at that as well.

@dd8 It is the W3C Slack. See [the wiki page](https://www.w3.org/wiki/Slack) for joining.

Jym77 committed:
Merge branch 'develop' into target-size-enhanced

Jym77 committed:
Improve definition

Jym77 committed:
Do not use aria-labelledby on raw div

Jym77 committed:
Add definitions of implicit and explicit labels

Jym77 committed:
Add definition of shape-shifted element

Jym77 committed:
Clean up

Jym77 committed:
Remove example numbering until solidification

Jym77 committed:
Merge branch 'develop' into target-size-enhanced

Jym77 committed:
Revert unrelated typo fix

Jym77 committed:
Write definitions

Jym77 committed:
Typos

Jym77 committed:
Revert unrelated typo fix

Jym77 committed:
Remove unrelated typo fix

Jym77 committed:
Merge branch 'develop' into target-size-enhanced

carlosapaduarte committed:
Merge branch 'develop' into target-size-enhanced

carlosapaduarte commented on _rules/target-size-enhanced-gi8qkf.md:
```diff
@@ -0,0 +1,731 @@
+For each test target, at least one of the following is true:
+
+- the target element has a [clickable area][] containing a rectangle aligned with the screen border whose width and height are at least 44 CSS pixels; or
+- though scrolling, the element can be brought into viewport with a [clickable area][] width and height of at least 44 CSS pixels; or
```
These are not  applicable given the current definition of 'Can be target by a pointer event' because it requires that 'the element's [border box][] intersects the viewport'

WilcoFiers replied:
This is a really neat idea! ❤️

carlosapaduarte commented on pages/glossary/can-be-targeted-by-pointer-event.md:
```diff
@@ -0,0 +1,37 @@
+- the element is a [semantic `widget`][semantic role]; and
+- the element is [focusable][]; and
+- the element's [border box][] intersects the viewport; and
+- the element's [border box][] is not entirely covered by the [border boxes][] of elements with greater `z-index` and `pointer-events: auto` CSS properties.
```
Want to consider also open dialogs?
https://html.spec.whatwg.org/multipage/interactive-elements.html#attr-dialog-open

Jym77 replied:
I'm not sure exactly what you mean 🤔 

Is it about elements "behind" the modal being blocked by it when it's opened (even if they are visible)?
I think that this is covered because in that case, their "clickable area" would be empty (the blocked elements are never topmost target, until the modal is closed and stops blocking them).

WilcoFiers commented on _rules/target-size-enhanced-gi8qkf.md:
```diff
@@ -0,0 +1,731 @@
+
+For each test target, at least one of the following is true:
+
+- the target element has a [clickable area][] containing a rectangle aligned with the screen border whose width and height are at least 44 CSS pixels; or
```
"Aligned with the screen border" feels off to me. Screens aren't necessarily rectangles. Some smart watches have circular screens for example. But this also seems like it would require some new input aspect of having to know the something about the edge of the screen?

WilcoFiers commented on _rules/target-size-enhanced-gi8qkf.md:
```diff
@@ -0,0 +1,731 @@
+
+- the target element has a [clickable area][] containing a rectangle aligned with the screen border whose width and height are at least 44 CSS pixels; or
+- though scrolling, the element can be brought into viewport with a [clickable area][] width and height of at least 44 CSS pixels; or
+- the element has an empty [clickable area][], and its [clickable area][] cannot be made non-empty through scrolling; or
+  > Note: this is mostly for totally covered or off-screen elements. This should probably rather be handled in the Applicabilty and make these Inapplicable, but I'm not sure it can be done with the current definition…
```
What is an "empty" clickable area?

Jym77 replied:
"clickable area" are defined as sets of pixels. So, it is the empty set.

WilcoFiers commented on _rules/target-size-enhanced-gi8qkf.md:
```diff
@@ -0,0 +1,731 @@
+  > Note: this is mostly for totally covered or off-screen elements. This should probably rather be handled in the Applicabilty and make these Inapplicable, but I'm not sure it can be done with the current definition…
+- the target is part of [inline text][]; or
+  > comment: I feel this is going to be difficult to define objectively, so it is better in Expectation for now. #1010 has some work in that direction that we can probably reuse: https://github.com/act-rules/act-rules.github.io/pull/1010/files#diff-32079a0602a5a909b242b4e0961e7c5ddd6b6f5c9906b216d5bf21cf2ba13a77R28-R29 > https://github.com/act-rules/act-rules.github.io/blob/4b64bba6cb77a8d4dc0649 ...
+- The size is [essential target size][]
+  > comment: this is always a bit tricky. I guess we can do as in #1916 and list cases that are considered essential (list can grow).
```
I would be all in favor of an assumption for this one. Do we know of real-world examples?

Jym77 replied:
I've added the definition.
This is mostly for pins on a map, and dots on a graph.

Jym77 committed:
Merge branch 'develop' into target-size-enhanced

Jym77 commented:
I've updated the rule with comments for review, and cleaned it a bit. It now starts to look like an actual rule 🎉 (still need numbering of test cases, and a few fixes on some).

* Ultimately, I've ditched the "weird size" exceptions. These are only really needed for the 2.5.8 rule (due to distance), and there was no real need to restrict this rule because of that. I made sure that the examples do not rely on the weird size (i.e., no 50×50 button clipped to 40×40), so that automated tools who rely on border boxes or `getBoundingClientRect` can still claim a consistent implementation (and added a note to explain that).
* Also simplified a bit some of the definitions…

Jym77 committed:
Update rule and definition

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/act-rules/act-rules.github.io?pullRequest=1919) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/act-rules/act-rules.github.io?pullRequest=1919) before we can accept your contribution.<br/><hr/>**Carlos Duarte** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/act-rules/act-rules.github.io?pullRequest=1919) it.</sub>

Jym77 committed:
Merge branch 'develop' into target-size-enhanced

Jym77 committed:
Clean up examples

Jym77 committed:
Move UA controlled exception from Applicability to Expectation

Jym77 committed:
Typo in link

WilcoFiers commented:
Superseded by #2279
